src/mavedb/lib/taxonomies.py

```diff
@@ -0,0 +1,61 @@
+import httpx
+from fastapi.exceptions import HTTPException
+from sqlalchemy.orm import Session
+from typing import Any
+
+from mavedb.models.taxonomy import Taxonomy
+
+
+async def find_or_create_taxonomy(db: Session, taxonomy: Taxonomy):
+    """
+    Find an existing taxonomy ID record with the specified tax_id int, or create a new one.
+
+    :param db: An active database session
+    :param tax_id: A valid taxonomy ID from NCBI
+    :return: An existing Taxonomy containing the specified taxonomy ID, or a new, unsaved Taxonomy
+    """
+    taxonomy_record = db.query(Taxonomy).filter(Taxonomy.tax_id == taxonomy.tax_id).one_or_none()
+    if not taxonomy_record:
+        taxonomy_record = await search_NCBI_taxonomy(db, taxonomy.tax_id)
+    return taxonomy_record
+
+
+async def search_NCBI_taxonomy(db: Session, search: str) -> Any:
+    """
+    Check whether this taxonomy exists in NCBI website.
+    If it exists, saving it in local database and return it to users.
+    """
+    if search and len(search.strip()) > 0:
+        if search.isnumeric() is False:
+            search_text = search.strip().lower()
+        else:
+            search_text = str(search)
+
+        url = f"https://api.ncbi.nlm.nih.gov/datasets/v2alpha/taxonomy/taxon/{search_text}"
+        async with httpx.AsyncClient() as client:
```

Comment:
when we fetch pubmed articles from NCBI, we don't use httpx but rather Eutils with our NCBI API key. I think you should probably use that here so we don't run into issues being rate limited from using the free api.

link: https://github.com/VariantEffect/mavedb-api/blob/89945b5310d2eacf1ce391398e3b8d298e5810e4/src/mavedb/lib/identifiers.py#L182

Reply:
I might need time to do some searches about it. Taxonomy seems in EInfo.

Reply:
- Eutils doesn't support taxonomy database. 
- EInfo only gets database information. 
- ESearch only supports term search. It neither supports tax_id search nor return all of necessary values we need. 
- Biopython Entrez.efetch only returns taxonomy's taxId and ScientificName. It's mainly used in getting GenBank data. 
- ELink can be used to find related items in the NCBI Entrez databases. Only getting related ID lists. I don't understand what these related IDs are for taxonomy. PubMed can get information that if a paper has been cited. 

NCBI doesn't appear to have a way to create an NCBI API key that is tied to "MaveDB" and not a person with a user account. - Alan

I think getting NCBI taxonomy in original way is fine. Not so many new taxonomies are found every year. It's different from lot of new publication papers published every year. The dropdown menu list covers the common taxonomies should work for most cases so usually MaveDB doesn't need to send a request to NCBI.